crates/rethnet_evm/src/db/fork.rs

```diff
@@ -0,0 +1,298 @@
+        let remote_db = RemoteDatabase::new(url);
+
+        let layered_db = LayeredDatabase::<RethnetLayer>::default();
+
+        Self {
+            layered_db,
+            remote_db,
+            account_info_cache: HashMap::new(),
+            code_by_hash_cache: HashMap::new(),
+            storage_cache: HashMap::new(),
+            fork_block_number,
+            fork_block_state_root_cache: None,
+        }
+    }
+}
+
+impl revm::Database for ForkDatabase {
+    type Error = ForkDatabaseError;
+
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        if let Some(layered) = self
+            .layered_db
+            .basic(address)
+            .map_err(ForkDatabaseError::LayeredDatabase)?
+        {
+            Ok(Some(layered))
+        } else if let Some(cached) = self.account_info_cache.get(&address) {
+            Ok(Some(cached.clone()))
+        } else if let Some(remote) = self
+            .remote_db
+            .basic(address)
+            .map_err(ForkDatabaseError::RemoteDatabase)?
+        {
+            self.account_info_cache.insert(address, remote.clone());
+
+            if remote.code.is_some() {
+                self.code_by_hash_cache
+                    .insert(remote.code_hash, remote.code.clone().unwrap());
+            }
+
+            Ok(Some(remote))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        if let Ok(layered) = self.layered_db.code_by_hash(code_hash) {
+            Ok(layered)
+        } else if let Some(cached) = self.code_by_hash_cache.get(&code_hash) {
+            Ok(cached.clone())
+        } else {
+            // remote_db doesn't support code_by_hash, so there's no delegation to it here.
+            Err(ForkDatabaseError::NoSuchCodeHash(code_hash))
+        }
+    }
+
+    fn storage(&mut self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        let layered = self
+            .layered_db
+            .storage(address, index)
+            .map_err(ForkDatabaseError::LayeredDatabase)?;
+
+        if layered != U256::from(0) {
+            Ok(layered)
+        } else if let Some(cached) = self.storage_cache.get(&(address, index)) {
+            Ok(*cached)
+        } else {
+            let remote = self
+                .remote_db
+                .storage(address, index)
+                .map_err(ForkDatabaseError::RemoteDatabase)?;
+
+            self.storage_cache.insert((address, index), remote);
+
+            Ok(remote)
+        }
+    }
+}
+
+impl revm::DatabaseCommit for ForkDatabase {
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
+        self.layered_db.commit(changes)
+    }
+}
+
+impl crate::DatabaseDebug for ForkDatabase {
+    type Error = ForkDatabaseError;
+
+    fn account_storage_root(&mut self, address: &Address) -> Result<Option<B256>, Self::Error> {
+        self.layered_db
```

Comment:
I just noticed that we're not handling cases where the account only exists in the fork block.

You can handle that by checking `if let Some(..)` and otherwise returning the storage root for the fork block.

Reply:
The reason I left this out is because there is no way to retrieve the storage _root_ via the JSON-RPC API. One can obtain the storage value at a particular index (via `eth_getStorageAt`), but from what I can tell there is no API for retrieving the storage root for an account (only the broader state root for a block).

Reply:
I just checked the fork state manager and that leaves it undefined as well, so indeed no need to implement it.